Simple write example

Hi,
I tried the simple write example: https://github.com/gimli-rs/gimli/blob/master/crates/examples/src/bin/simple_write.rs
I ran it, got the object file. 
Then when i tried to compile it into an executable i got following errors:
```
PS C:\Users\toni-\Desktop\gimli_test> gcc hello.o 
hello.o:hello.c:(.debug_line+0x22): relocation truncated to fit: IMAGE_REL_AMD64_ADDR32 against `.debug_line_str'
hello.o:hello.c:(.debug_line+0x2c): relocation truncated to fit: IMAGE_REL_AMD64_ADDR32 against `.debug_line_str'
hello.o:hello.c:(.debug_line+0x31): relocation truncated to fit: IMAGE_REL_AMD64_ADDR32 against `.debug_line_str'
hello.o:hello.c:(.debug_info+0x8): relocation truncated to fit: IMAGE_REL_AMD64_ADDR32 against `.debug_abbrev'
hello.o:hello.c:(.debug_info+0x31): relocation truncated to fit: IMAGE_REL_AMD64_ADDR32 against `.debug_rnglists'
hello.o:hello.c:(.debug_info+0x35): relocation truncated to fit: IMAGE_REL_AMD64_ADDR32 against `.debug_line'
collect2.exe: error: ld returned 1 exit status
```
So i gone ahed and changched the format from `Dwarf32` to `Dwarf64`. This time the linking actually worked. But then as i gone to test it in gdb. I got following error:
```
Dwarf Error: bad offset (0x140010000) in compilation unit header (offset 0x0 + 6) [in module C:\Users\toni-\Desktop\gimli_test\a.exe]
```

My gcc is:
```
gcc.exe (MinGW-W64 x86_64-msvcrt-posix-seh) 14.1.0
Copyright (C) 2024 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```
Bye

--- a/crates/examples/src/bin/simple_write.rs
+++ b/crates/examples/src/bin/simple_write.rs
@@ -3,7 +3,7 @@
 //! The resulting object file can be linked with a C runtime to create a complete executable:
 //! ```sh
 //! $ cargo run --bin simple_write
-//! $ gcc -o hello hello.o
+//! $ gcc -o hello hello.o -z noexecstack
 //! $ ./hello
 //! Hello, world!
 //! ```
@@ -48,8 +48,9 @@ impl RelocateWriter for Section {
 }
 
 fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let binary_format = object::BinaryFormat::native_object();
     let mut obj = object::write::Object::new(
-        object::BinaryFormat::native_object(),
+        binary_format,
         object::Architecture::X86_64,
         object::Endianness::Little,
     );
@@ -69,7 +70,12 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
     // Choose the encoding parameters.
     let encoding = Encoding {
         format: Format::Dwarf32,
-        version: 5,
+        version: if binary_format == object::BinaryFormat::Coff {
+            // The COFF toolchain I used didn't work with DWARF version 5.
+            4
+        } else {
+            5
+        },
         address_size: 8,
     };
 
@@ -148,7 +154,12 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
         if section.data.len() == 0 {
             return Ok(());
         }
-        let section_id = obj.add_section(Vec::new(), id.name().into(), object::SectionKind::Debug);
+        let kind = if id.is_string() {
+            object::SectionKind::DebugString
+        } else {
+            object::SectionKind::Debug
+        };
+        let section_id = obj.add_section(Vec::new(), id.name().into(), kind);
         obj.set_section_data(section_id, section.data.take(), 1);
 
         // Record the section ID so that it can be used for relocations.
@@ -166,14 +177,20 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
             // The `eh_pe` field is not used in this example because we are not writing
             // unwind information.
             debug_assert!(reloc.eh_pe.is_none());
-            let symbol = match reloc.target {
+            let (symbol, kind) = match reloc.target {
                 RelocationTarget::Section(id) => {
-                    obj.section_symbol(sections.get(id).unwrap().id.unwrap())
+                    let kind = if binary_format == object::BinaryFormat::Coff {
+                        object::RelocationKind::SectionOffset
+                    } else {
+                        object::RelocationKind::Absolute
+                    };
+                    let symbol = obj.section_symbol(sections.get(id).unwrap().id.unwrap());
+                    (symbol, kind)
                 }
                 RelocationTarget::Symbol(id) => {
                     // The main function is the only symbol we have defined.
                     debug_assert_eq!(id, 0);
-                    main_symbol
+                    (main_symbol, object::RelocationKind::Absolute)
                 }
             };
             obj.add_relocation(
@@ -183,7 +200,7 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
                     symbol,
                     addend: reloc.addend,
                     flags: object::RelocationFlags::Generic {
-                        kind: object::RelocationKind::Absolute,
+                        kind,
                         encoding: object::RelocationEncoding::Generic,
                         size: reloc.size * 8,
                     },

--- a/src/common.rs
+++ b/src/common.rs
@@ -365,6 +365,13 @@ impl SectionId {
             _ => return None,
         })
     }
+
+    /// Returns true if this is a mergeable string section.
+    ///
+    /// This is useful for determining the correct section flags.
+    pub fn is_string(self) -> bool {
+        matches!(self, SectionId::DebugStr | SectionId::DebugLineStr)
+    }
 }
 
 /// An optionally-provided implementation-defined compilation unit ID to enable

